File: java/347.top-k-frequent-elements.java
/*
 * @lc app=leetcode id=347 lang=java
 *
 * [347] Top K Frequent Elements
 */
class Solution {
    public List<Integer> topKFrequent(int[] nums, int k) {
        List<Integer> res = new LinkedList<>();
        if (nums == null || nums.length == 0) return res;
        Map<Integer, Integer> cntMap = new HashMap<>();
        for (int num : nums) {
            cntMap.put(num, cntMap.getOrDefault(num, 0) + 1);
        }
        PriorityQueue<CntPair> queue = new PriorityQueue<>((p1, p2) -> {
            if (p1.cnt < p2.cnt) return 1;
            else if (p1.cnt > p2.cnt) return -1;
            else return 0;
        });
        for (Map.Entry<Integer, Integer> entry : cntMap.entrySet()) {
            queue.offer(new CntPair(entry.getKey(), entry.getValue()));
        }
        while (!queue.isEmpty() && k-- > 0) {
            res.add(queue.poll().num);
        }
        return res;
    }

    static class CntPair {
        int num;
        int cnt;
        public CntPair(int num, int cnt) {
            this.num = num;
            this.cnt = cnt;
        }
    }
}

File: go/347.top-k-frequent-elements.go
/*
 * @lc app=leetcode id=347 lang=golang
 *
 * [347] Top K Frequent Elements
 */

// @lc code=start
func topKFrequent(nums []int, k int) []int {
	res := make([]int, 0)
	freq := make(map[int]int)
	for _, num := range nums {
		freq[num]++
	}
	cnt := make(map[int][]int)
	cntList := make([]int, len(freq))
	i := 0
	for k, v := range freq {
		cnt[v] = append(cnt[v], k)
		cntList[i] = v
		i++
	}
	i = i-1
	sort.Ints(cntList)
	for k > 0 {
		items := cnt[cntList[i]]
		res = append(res, items...)
		i, k = i-len(items), k-len(items)
	}
	return res
}
// @lc code=end

